### PR TITLE
[Integration][New Relic] - Add support for service level

### PR DESCRIPTION
**Pull request by PeyGis**
# Description

What - Added support for New Relic Service Levels
Why -
How - Used the [NerdGraphQL API](https://docs.newrelic.com/docs/apis/nerdgraph/get-started/introduction-new-relic-nerdgraph/) to query entities of type SERVICE_LEVEL

## Type of change

Please leave one option from the following and delete the rest:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] New Integration (non-breaking change which adds a new integration)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Non-breaking change (fix of existing functionality that will not change current behavior)
- [ ] Documentation (added/updated documentation)

## Screenshots

Screenshot 2024-07-24 at 9 13 06 PM

## API Documentation

Sample response
```json
{
   "serviceLevel":{
      "indicators":[
         {
            "createdAt":1721030473712,
            "createdBy":{
               "email":"stav@getport.io"
            },
            "description":"Proportion of requests that are served without errors.",
            "guid":"NDM2OTY4MHxFWFR8U0VSVklDRV9MRVZFTHw1OTk0MzM",
            "id":"599433",
            "name":"Port API Stg 01 - Success",
            "objectives":[
               {
                  "description":"None",
                  "name":"None",
                  "target":98.47,
                  "timeWindow":{
                     "rolling":{
                        "count":7,
                        "unit":"DAY"
                     }
                  }
               }
            ],
            "resultQueries":{
               "indicator":{
                  "nrql":"SELECT clamp_max((sum(newrelic.sli.valid) - sum(newrelic.sli.bad)) / sum(newrelic.sli.valid) * 100, 100) AS 'SLI' FROM Metric WHERE entity.guid = 'NDM2OTY4MHxFWFR8U0VSVklDRV9MRVZFTHw1OTk0MzM' UNTIL 2 minutes AGO"
               }
            },
            "updatedAt":"None",
            "updatedBy":"None"
         }
      ]
   },
   "tags":{
      "account":[
         "Account 436968[REDACTED]"
      ],
      "accountId":[
         "436968[REDACTED]"
      ],
      "category":[
         "success"
      ],
      "nr.associatedEntityGuid":[
         "NDM2OTY4MHxBUE18QVBQTElDQVRJT058NTkxMTYyMjE0"
      ],
      "nr.associatedEntityName":[
         "Port API Stg 01"
      ],
      "nr.associatedEntityType":[
         "APM_APPLICATION"
      ],
      "nr.sliComplianceCategory":[
         "Compliant"
      ],
      "nr.sloPeriod":[
         "7d"
      ],
      "nr.sloTarget":[
         "98.47%"
      ],
      "trustedAccountId":[
         "436968[REDACTED]"
      ]
   },
   "__SLI":{
      "SLI":99.92846924177397
   }
}
```

### CODE REVIEW

File: integrations/newrelic/newrelic_integration/core/query_templates/service_levels.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+GET_SLI_BY_NRQL_QUERY = """
+{
+  actor {
+    account(id: {{ account_id }}) {
+      nrql(query: "{{ nrql_query }}") {
+  results
```
[the remark this answers is not in the document]
Comment (PeyGis) [reply integrations/newrelic/newrelic_integration/core/query_templates/service_levels.py:6]:
```json
"indicator":{
                  "nrql":"SELECT clamp_max((sum(newrelic.sli.valid) - sum(newrelic.sli.bad)) / sum(newrelic.sli.valid) * 100, 100) AS 'SLI' FROM Metric WHERE entity.guid = 'NDM2OTY4MHxFWFR8U0VSVklDRV9MRVZFTHw1OTk0MzM' UNTIL 2 minutes AGO"
               }
```

File: integrations/newrelic/newrelic_integration/core/service_levels.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+        async for service_level in send_paginated_graph_api_request(
+            self.http_client,
+            LIST_SLOS_QUERY,
+            request_type="list_service_levels",
+            extract_data=self._extract_service_levels,
+        ):
+            nrql = (
+                service_level.get("serviceLevel", {})
+                .get("indicators", [])[0]
```
[the remark this answers is not in the document]
Comment (PeyGis) [reply integrations/newrelic/newrelic_integration/core/service_levels.py:53]: I confirmed this and it is not possible to have more than 1

File: integrations/newrelic/newrelic_integration/core/service_levels.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            .get("results", [])
+        )
+        if service_levels:
+            return service_levels[0]
+        return {}
+
+    async def list_service_levels(self) -> AsyncIterable[dict[str, Any]]:
+
+        async for service_level in send_paginated_graph_api_request(
+            self.http_client,
```
Comment (Tankilevitch) [inline integrations/newrelic/newrelic_integration/core/service_levels.py:46]: if you look inside the `send_paginated_graph_api_request` it actually already fetched a batch, but returns it item by item, so to make it return faster, and maybe async gather the SLIs for multiple SLOs I would suggest either adjusting the `send_paginated_graph_api_request` or gathering a size of a batch and then bringing the `SLI` for each one, and using the `stream_async_iterators_tasks` to return the results for each one.

File: integrations/newrelic/newrelic_integration/core/service_levels.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            LIST_SLOS_QUERY,
+            request_type="list_service_levels",
+            extract_data=self._extract_service_levels,
+        ):
+            nrql = (
+                service_level.get("serviceLevel", {})
```
Comment (Tankilevitch) [inline integrations/newrelic/newrelic_integration/core/service_levels.py:52]: add comment about that you are taking the `nrql` that the SLI is actually built from and query it to get the result

**Review by Tankilevitch — APPROVED** [review]

File: integrations/newrelic/CHANGELOG.md [hunk 5]
```diff
@@ -7,14 +7,54 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 <!-- towncrier release notes start -->
 
-# Port_Ocean 0.1.65 (2024-08-01)
+# Port_Ocean 0.1.70 (2024-08-15)
```
Comment (Tankilevitch) [inline integrations/newrelic/CHANGELOG.md:10]:
```
## 0.1.70 (2024-08-15)
```

**Review by Tankilevitch — APPROVED** [review]

**Pull request merged**